File: ios/RNMBX/ImageManager.swift
import MapboxMaps
/**
  ImageManager helps to resolve images defined by any of RNMBXImages component.
 */
class ImageManager {
  typealias Resolver = (String, UIImage) -> Void
  var subscriptions: [String: [Subscription]] = [:]
  
  class Subscription : Cancelable {
    var name: String
    var resolved: Resolver
    weak var manager: ImageManager?
    
    init(name: String, resolved: @escaping Resolver) {
      self.name = name
      self.resolved = resolved
    }
    
    func cancel() {
      manager?.unsubscript(subscription: self)
    }
  }
  
  func subscribe(name: String, resolved: @escaping Resolver) -> Subscription {
    var subscription = Subscription(name: name, resolved: resolved)
    var list = subscriptions[name] ?? []
    list.append(subscription)
    subscriptions[name] = list
    return subscription
  }
  
  func unsubscript(subscription: Subscription) {
    var list = subscriptions[subscription.name] ?? []
    list.removeAll { $0 === subscription }
    subscriptions[subscription.name] = list
  }

  func resolve(name: String, image: UIImage) {
    subscriptions[name]?.forEach { $0.resolved(name, image) }
  }
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/images/ImageManager.kt
package com.rnmapbox.rnmbx.components.images

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import com.rnmapbox.rnmbx.v11compat.Cancelable
import com.mapbox.maps.Image
import com.rnmapbox.rnmbx.v11compat.image.toMapboxImage

/**
ImageManager helps to resolve images defined by any of RNMBXImages component.
 */

fun interface Resolver {
    fun resolved(name: String, image: Image)
}
class Subscription(val name:String, val resolver: Resolver, val manager: ImageManager): Cancelable {

    fun resolved(name: String, image: Image) {
        resolver.resolved(name, image)
    }
    override fun cancel() {
        manager.unsubscribe(this)
    }
}

class ImageManager {
    var subscriptions: MutableMap<String, MutableList<Subscription>> = hashMapOf()

    fun subscribe(name: String, resolved: Resolver) : Subscription {
        val list = subscriptions.getOrPut(name) { mutableListOf() }
        val result = Subscription(name, resolved, this)
        list.add(result)
        return result
    }
    fun unsubscribe(subscription: Subscription) {
        var list = subscriptions[subscription.name]
        list?.removeAll { it === subscription }
    }

    fun resolve(name: String, image: Image) {
        subscriptions[name]?.forEach {
            it.resolved(name, image)
        }
    }

    fun resolve(name: String, image: Bitmap) {
        resolve(name, image.toMapboxImage())
    }

    fun resolve(name: String, image: BitmapDrawable) {
        resolve(name, image.bitmap)
    }
}